Implement differentiable CGP

jakobj opened the pull request:
- add a new node CGPParameter that, in contrast to the ConstantFloat can store arbitrary values
- these parameter values are stored in each individual and passed on to their offspring
- parameters can be updated from torch classes generated from the graph; this way parameters can be optimized with local search and then written back to the individuals
- add a new function template to provide local search via torch optimizers

as discussed. rebased onto master and added some improvements. this supersedes #57.

closes #59

jakobj commented:
i introduced the gradient clipping (https://github.com/jakobj/python-gp/blob/57299e9a9b34ab9dd2ac6642ddea064f5b672cb5/gp/local_search/gradient_based.py#L50) since i've experienced a few issues with exploding gradients. however i think this makes sense in general: as the parameters are passed on to offspring we do not want them to be tuned to ridiculous values if the graph is completely wrong. instead we want them to be adapted in parallel to the graph over many generations.
i am not sure whether we want to include this explicitly like this or let users include it if they need it. this would require making the `gradient_based` function a bit more flexible. what do you think @mschmidt87 ?

jakobj commented:
there's another interesting problem: the current implementation relies on `Parameter` nodes that provide numeric values to the graph. due to the logic of CGP these nodes can be used more than once in the final expression. for example it can happen that a node is used with opposite signs  in an expression as in `f(x) = -c * x** 2 + c` if the final expression is something like `f(x) = -3 * x ** 2 + 1` and the current value of c is 2.0 the gradient will pull in opposing directions, essentially halting the search although the structure of the expression is already correct. other approaches, e.g., Izzo et al 2017, consider multiplicative constants in each node instead. these are private to each node an hence can not lead to this effect.
one could consider an implementation where each node has a multiplicative and an additive constant that by default are one and zero, that could be tuned via local search. this would make the `Parameter` node proposed here obsolete. what do you think @mschmidt87 ?

mschmidt87 commented on examples/example_differential_evo_regression.py:
```diff
@@ -0,0 +1,138 @@
+    history_fitness = np.array(history["fitness_parents"])
+    ax_fitness.plot(np.max(history_fitness, axis=1), label="Champion")
+    ax_fitness.plot(np.mean(history_fitness, axis=1), label="Population mean")
+
```
```suggestion
ax_fitness.set_yscale('symlog')
```

jakobj commented on gp/individual.py:
```diff
@@ -129,7 +149,33 @@ def to_sympy(self, simplify=True):
+
+        """
+        for name, value in torch_cls.named_parameters():
+            name = f"<{name[1:]}>"
```
nothing, they are meant literally as in the string should be something like the string "<p1>". we also cut of the leading underscore "_p1" -> "<p1">

mschmidt87 replied:
Got it.

jakobj commented on gp/local_search/gradient_based.py:
```diff
@@ -0,0 +1,56 @@
+        raise ModuleNotFoundError("No module named 'torch' (extra requirement)")
+
+    if optimizer is None:
+        optimizer_class = torch.optim.SGD
```
i knew i should've commented on this ;) we can't do that bc we first need to check in line 31 whether torch is available. otherwise it would make sense as you suggest.

jakobj replied:
sure, we could. i just handled it like this bc we do it similarly in other parts, e.g., the tests. i guess the question is whether this remains the only function in this module or not. also, i'm not sure when to raise the error if we put the check in the `__init__`? when the users tries to access `gradient_based`?

mschmidt87 replied:
I mean, if `torch` is not available, this module should not be available. If you do something like this in `__init__.py` this should work, right?
```python
try:
    import torch  # noqa: F401
    from .gradient_based import gradient_based
except ModuleNotFoundError:
    pass

jakobj replied:
yes, of course. but how do we handle the case where torch is not available and the user tries to `local_search.gradient_based`. they would just get a noninformative "ModuleNotFoundError", right? the way it's currently implemented they will have the information that this exists, but they need the extra requirement to use it.

mschmidt87 replied:
Okay, I see your point. What about doing this then:
```python

try:
    import torch
except ModuleNotFoundError:
    torch = None

if torch is None:
    def gradient_based(*args, **kwargs):
          raise ModuleNotFoundError("No module named 'torch' (extra requirement)")
else:
    def gradient_based(..., optimizer=torch.optim.SGD, ...):
        ## full function definition
```

jakobj replied:
interesting, haven't thought of that :thinking: do you think this is easier to understand than just accepting a `None` as a default? or are you afraid this is going to lead to problems with mypy? besides the seeemingly increased complexity, the additional indentation bugs me

mschmidt87 replied:
I think it is easier to understand, but I get the point with the additional indent. Maybe it's fine to leave it as it is.

jakobj commented:
thanks for the review! i've addressed your comments. i've also made some minor changes in the parameters and introduced a new default for `clip_values`. please have another look.

mschmidt87 commented:
> one could consider an implementation where each node has a multiplicative and an additive constant that by default are one and zero, that could be tuned via local search. this would make the `Parameter` node proposed here obsolete. what do you think @mschmidt87 ?

I like that suggestion. However, what about first merging this and then making this a feature of the next release? We could add a warning in the documentation for the user to raise awareness of the problem with the gradients? Or do you think this is a deal-breaker?

jakobj commented:
agreed, no need to solve everything in one PR. the current implementation is good for now, but yes, it has some algorithmic problems. let's address those in a later version. i don't think we need documentation on this. this sounds more like research than documentation ;)

mschmidt87 commented:
Sounds good to me.

mschmidt87 commented:
As far as I can see, just the one point about availability of torch is remaining.

But one more thing: Can you add some description to the README about this?

mschmidt87 commented:
I'm fine with the code changes, just add some documentation to the README please.

jakobj commented:
ok, i've updated the README, please have another look @mschmidt87

mschmidt87 commented on README.md:
```diff
@@ -13,96 +15,75 @@ This library implements Cartesian Genetic Programming (e.g, Miller and Thomson,
+Searching numerical values for constants via local search
+---------------------------------------------------------
+
+This library supports the use of constants as operator nodes in order to represent expressions like `f(x) = x + 1.0`.
```
better use `c` in the equation instead of `1.0`.

jakobj commented:
good point, fixed it now. if you're happy pls approve so that i can merge ;)

jakobj commented:
travis and you are happy. using my administrator power to merge :smiling_imp:
